AfterClass-io/afterclass.io-v2
# feat: search api integration with postgres fulltext search

closes #140 

## Context



previously, there was effort made to introduce search UI at #152. this PR completes that line of work by introducing full searching and filtering search results functionality

## Changes


- abstract subcomponents of `<SearchResult/>`
- update `<SearchResultContent/>` to encapsulate search filtering logic
- update `<searchResult.Title/>` component to be server-sided
- expose radix-ui toggle group's `onValueChange` prop at `<SearchResult.Filter/>`
- add search function `searchProf` and `searchCourse` at `@/common/functions/*`


## Implementation Details




Search
- using [postgres's full text search](https://www.postgresql.org/docs/current/textsearch-intro.html)
- make query using [prisma's $queryRaw](https://www.prisma.io/docs/orm/prisma-client/queries/raw-database-access/raw-queries#queryraw) api

> [!IMPORTANT]
> note that the quick and dirty sql query used is meant to be temporary, until we can find a sensible search solution with #72 

Filter
- radix-ui toggle group with exposed hook to update filter state
- conditional rendering based on filter state
- ranking: course is always ranked above professor

> [!WARNING]
> the review count & course/professor count statistics icon is currently using hard coded values (11 & 31) this will be addressed as a separate task at #162 

## How to Test


1. head to preview link
2. click `/` to open search bar
3. search anything, search "L" for example
4. should see results
    a. if no results, should see a message "no results found" 
5. filtering should work


## Preview / Screenshots




![screenshot_20240612_112047](https://github.com/AfterClass-io/afterclass.io-v2/assets/13061926/70d7a694-e701-4313-a701-5d1e22e1c4a7)

## Checklist
